### PR TITLE
[v0.26.0rc][Backport][Refactor][DSA] Move TP weight switching into quantization methods

### PR DESCRIPTION
**Pull request by GDzhu01**
## What this PR does

Backports #12982 to `releases/v0.26.0rc`.

This moves tensor-parallel weight split, gather, and TP/full-weight switching into quantization methods through `TPWeightSwitchMixin` and declarative gather/repeat specifications. It enables the DSA-CP `o_proj` path to support the quantization layouts that explicitly declare TP weight-switch capability.

## v0.26.0rc adaptations

- Preserved the release branch's `get_moe_num_logical_experts` integration in the affected quantization schemes.
- Preserved the release branch's MXFP compatibility checks and newer MoE execution paths.
- Added the `Any` typing import required by the backported DSA-CP full-weight buffer annotation.

## Validation

- `git diff --check origin/releases/v0.26.0rc...HEAD`
- `ruff check` with the repository-pinned Ruff 0.14.0: passed
- `ruff format --check` with Ruff 0.14.0: passed

NPU-dependent unit and model tests were not run locally because this workspace is on a PC without `torch_npu`. The original change was functionally validated on Ascend with DeepSeek-V4-Flash W4A8 and W8A8-MTP as documented in #12982; release-branch CI is expected to provide backport-specific test coverage.

- vLLM version: v0.26.0
- vLLM main: https://github.com/vllm-project/vllm/commit/d02df748bf9efd99022f1a062597dc3cb3808485

### CODE REVIEW

File: vllm_ascend/quantization/tp_weight_switch.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+                    f"Invalid TP gather dim {gather_spec.gather_dim} for attr {gather_spec.attr_name!r} "
+                    f"with shape {tuple(tensor.shape)}."
+                )
+
+            tp_tensor = tensor.clone().detach().contiguous() if clone_tp_tensors else tensor.detach()
+            if clone_tp_tensors:
+                with torch.no_grad():
+                    tensor.set_(tp_tensor)
+            full_shape_list = list(tp_tensor.shape)
+            full_shape_list[dim] *= tp_size
```
Comment (GDzhu01) [inline vllm_ascend/quantization/tp_weight_switch.py:192]: 【可靠性】enable_tp_weight_switch 同样未校验 tp_size 为正，也没有确认 layer 的 full/local 尺寸比确实等于 tp_size。配置不一致时仍会分配一个错误大小的 full buffer，直到 collective 或内核阶段才失败。建议从 input/output size 推导并核对 TP 比例。

File: vllm_ascend/quantization/tp_weight_switch.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+                output=part.gather_output,
+                async_op=async_op,
+            )
+            if handle is not None:
+                state.handles.append(handle)
+
+    @staticmethod
+    def wait_tp_weight_all_gather(state: TPWeightSwitchState) -> None:
+        try:
+            for handle in state.handles:
```
Comment (GDzhu01) [inline vllm_ascend/quantization/tp_weight_switch.py:276]: 【可靠性】某个 handle.wait() 抛错后 finally 会直接清空全部 handles，后续尚未等待的 collective 仍可能在后台访问缓冲区，且状态失去追踪。建议尽力等待所有 handle、汇总首个异常后再清理，或只移除已完成项。

File: vllm_ascend/quantization/tp_weight_switch.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    @staticmethod
+    def split_tensor_for_tp(
+        tensor: torch.Tensor,
+        tp_size: int,
+        tp_rank: int,
+        dim: int = 0,
+        contiguous: bool = True,
+    ) -> torch.Tensor:
+        """Slice one TP shard from a full tensor."""
+        dim = dim if dim >= 0 else tensor.dim() + dim
```
Comment (GDzhu01) [inline vllm_ascend/quantization/tp_weight_switch.py:117]: 【安全、可靠性】split_tensor_for_tp 在索引 shape 前没有校验 tp_size>0、0<=tp_rank<tp_size 以及 dim 归一化后仍在范围内。非法值会表现为除零、IndexError，或 narrow 到错误分片。建议集中做参数校验并返回包含 shape/rank/size/dim 的明确异常。

File: vllm_ascend/quantization/tp_weight_switch.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+    def switch_tp_weight(
+        self,
+        layer: torch.nn.Module,
+        state: TPWeightSwitchState,
+        *,
+        use_full_weight: bool,
+    ) -> None:
+        """Switch layer tensor attributes between TP-local and full views."""
+        for attr_name, gather_part in state.gather_parts.items():
```
Comment (GDzhu01) [inline vllm_ascend/quantization/tp_weight_switch.py:289]: 【可靠性】switch_tp_weight 允许在 state.handles 仍非空时切到 full_tensor；异步 all-gather 尚未完成便可能被计算内核读取，结果存在数据竞争。建议 use_full_weight=True 时强制 handles 为空，或由该方法内部等待完成。

File: vllm_ascend/quantization/tp_weight_switch.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+                    tensor.set_(tp_tensor)
+            full_shape_list = list(tp_tensor.shape)
+            full_shape_list[dim] *= tp_size
+            full_shape = tuple(full_shape_list)
+            if dim == 0:
+                gather_input = tp_tensor
+            else:
+                gather_input = torch.movedim(tp_tensor, dim, 0).contiguous()
+            gather_shape = (full_shape[dim], *full_shape[:dim], *full_shape[dim + 1 :])
+            pool_key = (
```
Comment (GDzhu01) [inline vllm_ascend/quantization/tp_weight_switch.py:199]: 【可靠性、可维护性】使用共享 pool 且 pool_key_prefix=None 时，不同层只要 attr/shape/device/dtype 相同就会拿到同一 gather_output。若层间 gather 有重叠或权重切换生命周期交错，会互相覆盖 full weight。建议使用 pool 时强制唯一 layer prefix，或把层身份纳入 key。
